BoxContentSDK/BoxContentSDK/BOXNSURLSessionManager.h

```diff
@@ -10,21 +10,102 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+@protocol BOXNSURLSessionTaskDelegate <NSObject>
+
+/**
+ * To be called to finish the operation for a NSURLSessionTask upon its completion
+ *
+ * @param response  The response received from Box as a result of the API call.
+ * @param error     An error in the NSURLErrorDomain
+ */
+- (void)finishURLSessionTaskWithResponse:(NSURLResponse *)response error:(NSError *)error;
+
+@optional
+
+/**
+ * To be called to process the intermediate response from the task
+ *
+ * @param response  The intermediate response received from Box as a result of the API call
+ */
+- (void)processIntermediateResponse:(NSURLResponse *)response;
+
+/**
+ * To be called to process the intermediate data from the task
+ *
+ * @param data  The intermediate data received from Box as a result of the API call
+ */
+- (void)processIntermediateData:(NSData *)data;
+
+@end
+
+@protocol BOXNSURLSessionDownloadTaskDelegate <BOXNSURLSessionTaskDelegate>
+
+@optional
+
+/**
+ * Output stream to write data into. If destinationPath is provided, outputStream will be ignored
+ * Using outputStream to consume data will not allow the request to be executed in the background if the app is killed/suspended
+ */
+- (NSOutputStream *)outputStream;
+
+/**
+ * The location for output file. If provided, outputStream will be ignored
+ * Using destinationPath to consume data will allow request to be executed in the background
+ * if the app is killed/suspended and resume upon app restarts/resumes
+ * If download finishes successfully, i.e. at the time finishURLSessionTaskWithResponse:error:
+ *   method gets called, if no error, the downloaded file will be available at destinationPath
+ * The task delegate will be responsible for the downloaded file onwards
+ */
+- (NSString *)destinationPath;
```

Comment:
per your question about how to propage error if we fail to move the file in session manager, I think it's better to have an explicit method instead of destinationPath like - (void)didFinishDownloadingToURL:(NSURL *)location to notify about downloaded file exists and let the task delegate handle accordingly and handle error if any.